robot/web/web_c.go

```diff
@@ -375,14 +417,55 @@ func (svc *webService) handleVisualizeResourceGraph(w http.ResponseWriter, r *ht
 		}
 	}()
 
-	graph, err := graphviz.ParseBytes([]byte(dot))
+	graph, err := graphviz.ParseBytes([]byte(snapshot.Dot))
 	if err != nil {
 		return
 	}
 	if layout != "" {
 		gv.SetLayout(graphviz.Layout(layout))
 	}
-	if err = gv.Render(graph, graphviz.SVG, w); err != nil {
+
+	navButton := func(index int, label string) {
+		url := *r.URL
+		q := r.URL.Query()
+		q.Set(lookupParam, strconv.Itoa(index))
+		url.RawQuery = q.Encode()
+		var html string
+		if index < 0 || index >= snapshot.Count || index == lookup {
+			html = fmt.Sprintf(`<a>%s</a>`, label)
+		} else {
+			html = fmt.Sprintf(`<a href=%q>%s</a>`, url.String(), label)
+		}
+		write(html)
+	}
+
+	write(`<html><div>`)
+	navButton(0, "Latest")
+	write(`|`)
+	navButton(snapshot.Index-1, "Later")
+	// Index counts from 0, but we want to show pages starting from 1
+	write(fmt.Sprintf(`| %d / %d |`, snapshot.Index+1, snapshot.Count))
+	navButton(snapshot.Index+1, "Earlier")
+	write(`|`)
+	navButton(snapshot.Count-1, "Earliest")
+	write(`</div>`)
+
+	fxml := filterXML{w: w}
+	if err = gv.Render(graph, graphviz.SVG, fxml); err != nil {
```

Comment:
Double-checking, but I expect you've observed fxml.Write is getting called by `Render` exactly once? If it's being called multiple times, would this filtering be correct (i.e: did you observe multiple calls to `Write`, with each call being prefixed with unnecessary XML)?

[Commentary]
I think what you're observing is it that `gv.Render` adds some extra stuff XML junk in the middle of our nice HTML.

I think we can pull the comment from line 467 up here -- the "hack" isn't in the implemention of the type, but rather our call to `Render` is creating a special type to deal with our `Render` works.

And I think there's a couple choices in how to message this:
* I think we can document the behavior you observed. E.g: `Maxim: I observe Render calls Write a single time prefixed with 6 lines of XML we need to throw out.`
* or the behavior we depend on. E.g: `We depend on gv.Render calling fxml.Write exactly one time, and the first 6 lines are XML we do not want to pass to the final HTML`.

Reply:
yeah, that's a good point. will update and document

Reply:
yes `Render` calls `Write` once, but this could change so the second comment is probably more appropriate. 

the non-hack solution to identify the junk and remove it through a parser https://pkg.go.dev/golang.org/x/net/html, so I'll add a TODO for that as part of the explainer.